### PR TITLE
Version Vector with Exceptions

### PR DESCRIPTION
**Pull request by xldenis**
Here is an implementation of a causality buffer based on the algorithm presented in 

D. Malkhi and D. Terry, “Concise Version Vectors in WinFS,” p. 13.

The idea is to ensure we re-order messages so that they respect a causal order, which is important for datastructures like `LSEQ` where we need to receive each peer's messages in order but don't want to pay the cost of sending a full vector clock with each operation. 

The Causality Barrier is composed of two components: a version vector (with exceptions) and an unapplied operations buffer.

The Version Vector with Exceptions works by tracking two pieces of information for every known peer: the maximum message timestamp and a list of exceptions. The list of exceptions is composed of the timestamps less than the maximum which we haven't seen yet. 

When we receive a new operation we follow the following algorithm in pseudo-pseudo-code:

0. Update the version vector with A's timestamp, if  A > max then we set max = A and we add the range [old max, A) to the exceptions. Otherwise if A is in the exception set, we remove it's timestamp.  
1. Check if the operation A must occur after another operation B
2. Check if we've seen B
  a. If yes, then allow A, and check the unapplied buffer for opeations that transitively depend on A
  b. If no, buffer A

### CODE REVIEW

File: src/vvwe.rs [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+#![allow(missing_docs)]
+
+use derive_more::{Add, From, Into};
+use std::collections::*;
+use serde::{Deserialize, Deserializer, self, Serialize};
+
+/// Temporary
+#[derive(PartialEq, Eq, Debug, Copy, Clone, Hash, PartialOrd, Ord, From, Deserialize, Serialize)]
+pub struct SiteId(
+    #[serde(deserialize_with = "deserialize_number_from_string")]
```
Comment (chpio) [inline src/vvwe.rs:17]: Why is this needed? shouldn't serde just deserialize it the right way?

Comment (xldenis) [reply src/vvwe.rs:17]: https://github.com/serde-rs/json/issues/317

I need to check if it's still required here but the problem I was having was that the number would get turned into a stringified number by serde / js and then break deserialization :/

Comment (chpio) [reply src/vvwe.rs:17]: So it's some sort of compatibility code? For the case that the value was serialized as as string?

Comment (xldenis) [reply src/vvwe.rs:17]: Yea, it's been a while since I originally wrote that but I recall it having to do with serde generating weird instances for newtyped numbers.

Comment (xldenis) [reply src/vvwe.rs:17]: I'm planning on getting rid of the siteid entirely in favor of the more generic `Actor` trait

File: src/vvwe.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        // 2. Find exceptions that we've seen.
+        let mut local_exceptions  = other.exceptions.difference(&self.exceptions).map(|ix| ix.to_owned());
+
+        local_ops.chain(&mut local_exceptions).collect()
+    }
+}
+
+use crate::vclock::{Actor, Dot};
+
+pub trait CausalOp<A> {
```
Comment (xldenis) [inline src/vvwe.rs:72]: implementing this trait for different kinds of `Op` types allows this barrier to work with multiple CRDTs

File: src/vvwe.rs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        }
+    }
+
+    pub fn expel(&mut self, msg: T) -> T {
+        let v = self.peers.entry(msg.dot().actor).or_insert_with(VectorEntry::new);
+        v.increment(msg.dot().counter);
+        msg
+    }
+
+    pub fn diff_from(&self, other: &HashMap<A, VectorEntry>) -> HashMap<A, HashSet<LogTime>> {
```
Comment (xldenis) [inline src/vvwe.rs:136]: This is useful to generate "Anti-Entropy" messages which in turn gives you offline support and dynamic join/leave almost for free.

The basic idea is two actors exchange their version vectors, diff them, and gather all the messages they've seen locally that the other doesn't and send them to each other. After this operation both actors will have synchronized their states.

Comment (davidrusu) [reply src/vvwe.rs:136]: very cool

**Comment by davidrusu** [comment]
@xldenis I've been playing around with your branch and put up a PR on your fork with my changes as I explored it.

I think we can break apart the `VectorEntry` and bring it back into the CausalityBarrier because I feel it's hiding a VClock across those two layers.

here's a sketch:
``` rust
struct CauslityBarrier<A, T> {
    knowledge: VClock<A>
    exceptions: HashMap<Dot<A>, HashSet<T>> // collapse VectorEntry with the buffer
}
```

This should resolve our need to introduce the `VectorEntry` struct. thoughts?

File: src/vvwe.rs [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        } else {
+            let mut x = self.max_version + 1;
+            while x < clk {
+                self.exceptions.insert(x);
+                x = x + 1;
+            }
+        }
+    }
+
+    pub fn is_ready(&self, clk: &LogTime) -> bool {
```
Comment (bochaco) [inline src/vvwe.rs:51]: Perhaps something like `is_seen` is clearer

File: src/vvwe.rs [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+    pub fn increment(&mut self, clk: LogTime) {
+        // We've just found an exception
+        if clk < self.max_version {
+            self.exceptions.take(&clk);
+        } else if clk == self.max_version {
+            self.max_version = self.max_version + 1;
+        } else {
+            let mut x = self.max_version + 1;
+            while x < clk {
```
Comment (bochaco) [inline src/vvwe.rs:44]: A `for` loop would be easier to read I think

**Comment by xldenis** [comment]
> I think we can break apart the VectorEntry and bring it back into the CausalityBarrier because I feel it's hiding a VClock across those two layers.

That formulation is not equivalent and has worse memory behavior. 

Imagine a document where you see an edit from Actor 60000, with a vclock you're now going to allocate a vector of 60k elements to store the message counter. With the causality barrier you'll only insert one element. 

This is important if you want a document to gracefully handle add/dropping of users in an unbounded manner. The causality barrier only needs to store information about 'live' participants whereas the vector clock needs to store info about _all_ participants. In fact, one of the major points of the VVwE is to avoid a vector clock entirely!

**Comment by davidrusu** [comment]
> Imagine a document where you see an edit from Actor 60000, with a vclock you're now going to allocate a vector of 60k elements to store the message counter.

Take a look at the VClock implementation here: https://github.com/rust-crdt/rust-crdt/blob/master/src/vclock.rs#L38

The implementation stores a sparse vclock very similar to the CausalityBarrier you have here. (Actors with no explicit entry have an implicit counter of 0).

To expand a bit on the suggestion I wrote above:
``` rust
struct CauslityBarrier<A, T> {
    knowledge: VClock<A>
    exceptions: HashMap<Dot<A>, HashSet<T>> // collapse VectorEntry with the buffer
}
```
`knowledge` is the latest time for dots we've experienced in sequence.
`exceptions` is giving us a set of op's that are dependent on a particular `Dot<A>`.

Once we generalize the idea of "dependent operations" past just being dependent on a single dot, we'll have to change `exceptions` to be:

``` rust
    exceptions: HashMap<VClock<A>, HashSet<T>>
```

To allow the set of op's to be dependent on multiple operations.

**Comment by xldenis** [comment]
> The implementation stores a sparse vclock very similar to the CausalityBarrier you have here. (Actors with no explicit entry have an implicit counter of 0).

Very cool! In that case there should be no issue using a `VClock`. 

> Once we generalize the idea of "dependent operations" past just being dependent on a single dot, we'll have to change exceptions to be:

Are there any CRDTs currently in the library (or planned) where that's the case? Your generalization _seems_ correct but maybe there isn't yet a need for it?

**Comment by davidrusu** [comment]
Apologies for my slow comms, there's been some family things I've been focusing on lately.

> why we should be using a vclock?

Just wanted to share my intuition here so that we are on the same page. As much as we can, we want to rely on the primitives that have been tested and we understand well, introducing new ideas only when we absolutely need to. Also it simplifies the testing surface area a good deal.

> use cases for ops depending on multiple dots.

Good point, if it's not useful, then we shouldn't generalize. 

I have been working on a simpler Set CRDT that assumes causal delivery. If you take a close look at the ORSWOT that comes with this library, most of it's complexity comes from the `deferred_removes`. These are buffered member removals that are waiting for the corresponding insert to happen. This is essentially doing the work of a causality barrier you have here.

Now the hope is that a Set CRDT that externalizes this causality tracking can be internally much simpler.

Now to tie it back to multiple dots, we have ops in the ORSWOT that can remove multiple members at once, we'd want to have this sort of batched operation on the causal set as well.

If this works well, then naturally the Map CRDT can be simplified in the same way since it has the same internal structure as the ORSWOT.

File: src/vvwe.rs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    pub fn ingest(&mut self, op: T) -> Option<T> {
+        let v = self.peers.entry(op.dot().actor).or_default();
+        // Have we already seen this op?
+        if v.is_ready(op.dot().counter) {
+            return None;
+        }
+
+        v.increment(op.dot().counter);
```
Comment (davidrusu) [inline src/vvwe.rs:101]: I feel like we should be incrementing only after releasing the op. Otherwise we risk dropping ops from a source if we receive ops out of order.

Ie. imagine we receive two ops from site A, but we ingest the op with the more recent time before the older op. In this case, the older op will be dropped

**Comment by xldenis** [comment]
> Now to tie it back to multiple dots, we have ops in the ORSWOT that can remove multiple members at once, we'd want to have this sort of batched operation on the causal set as well.

Ah, yea that would definitely require a generalization then.

File: src/vvwe.rs (hunk 6 above)
Comment (xldenis) [reply src/vvwe.rs:101]: To clarify:

Assume we have:

OP1: (A, Clock1)
OP2: (A, Clock2)

with Clock2 < Clock1

Let's assume we have an empty VVwE
We process OP1 first. 

Our VVwE state will be { A: { next_version: Clock1 + 1, exceptions: [0 ... Clock1) } }

When we see OP2, the operation will not be 'ready' because it will be in the exception set. 

Assuming it doesn't depend on anything then we'll produce `Some(op)` on line 124.

**Comment by xldenis** [comment]
I was thinking about property tests and here are a couple I've thought of:

- Ingesting a set of insertions (independent ops) produces the same set of insertions regardless of order
- Generate an ordered sequence of operations (insertions _and_ deletes) and show that every permutation results in the same output (we'd need to hook it up to a crdt or something :/).
- Generate a bunch operations covering the range of clocks 0...N and show that whatever the order we ingest them in, once we've seen all of them the exception vector is empty. 

Of those obviously the second test is the most valuable but we need to make sure it is simple enough that we're confident the _test_ is correct. 

These could be generalized to be instructions from multiple actors but it would get even _more_ complex then. 

I attempted to create a `OperationList` arbitrary instance in the LSEQ PR. I feel like having a more general / powerful version of that would be a _very_ useful tool in general for this library.

File: src/vvwe.rs (hunk 6 above)
Comment (davidrusu) [reply src/vvwe.rs:101]: Ah thanks for expanding, I keep thinking in terms of vclocks, VectorEntry::increment does more than just increment!

Ok perhaps we should move to VectorEntry::ingest to clarify this

**Comment by davidrusu** [comment]
@xldenis thoughts on the comments here? https://github.com/xldenis/rust-crdt/pull/3#discussion_r414190019

This PR here https://github.com/xldenis/rust-crdt/pull/4 has a sketch to change the Barrier to buffer and release of ops as they become unblocked instead of cancelling them out. What do you think of that sketch there?

The next logical change would be to allow multiple ops to be waiting on a `dot`. Currently our buffer is just a `HashMap<Dot<A>, T>` we would need to move that to something like `HashMap<Dot<A>, Vec<T>>`.

Want to go ahead and make that change?

**Comment by davidrusu** [comment]
> re: testing

This is assuming something about the underlying crdt, in that inserts cancel out with deletes (this is why you break out the tests under `inserts` / `inserts + deletes`. I don't think we can assume this in general. Take for example the Map CRDT, the Update operation is much more complex than an insert, and it doesn't simply cancel out since it resets any deleted nested CRDTs

If we can agree to make the causal barrier only worry about releasing operations in causal order, than I think we can simplify the testing of this structure a fair amount.

Ie. we only need to ensure that operations are not released until after their happens_after constraint is satisfied.

This can be verified by just randomly generating a Vec<CausalOp>, passing it through the CausalityBarrier and ensuring that the vec we collect out on the other side is ordered such that any op's `happens_after()` is satisfied by ops earlier in the vec.

**Comment by xldenis** [comment]
> This can be verified by just randomly generating a Vec, passing it through the CausalityBarrier and ensuring that the vec we collect out on the other side is ordered such that any op's happens_after() is satisfied by ops earlier in the vec.

That's a good property!

**Comment by xldenis** [comment]
> Want to go ahead and make that change?

Yep, I'll start working on it. It'll be dependent on your PR that generalizes the return to `Vec<T>` because now the barrier will have to search the buffer for any transitive dependencies. After popping something from the buffer we could unblock other operations which are stuck.

**Comment by xldenis** [comment]
Rebased and squashed the commits @davidrusu
